python/ray/exceptions.py

```diff
@@ -350,6 +350,15 @@ def __str__(self):
         return error_msg
 
 
+@DeveloperAPI
+class UserCodeException(RayError):
+    """Indicates that an exception occurred while executing user code.
+    Used to catch user code exceptions which arise from executing
+    Ray Data code."""
```

Comment:
As it's a ray core level thing, better to say something more general here. for example, allow library developers to wrap user code exceptions from a remote task or actor. The `retry_exceptions` flag will still respect the cause of this exception.